### PR TITLE
feat: Implement friend system with relationship management

### LINKED ISSUE #174: Backend for friends management
### Overview
Implement the backend for the Friends feature. Users should be able to send, accept, decline, and cancel friend requests, as well as see their friends list and pending requests

### Constraints:
- Define data model(s) for friendships and friend requests (e.g. `PENDING`, `ACCEPTED`, `DECLINED`, `BLOCKED`)
- Expose endpoints for:
  - Sending a friend request.
  - Accepting / declining / cancelling a friend request
  - Listing current friends
  - Listing incoming and outgoing pending requests
- Add basic validation and permissions (cannot send request to yourself, cannot duplicate an existing pending/accepted request, etc)
- Ensure responses return all data the UI needs (ids, display name, avatar, status)
- Write basic tests for the main flows

(text reformulated by ChatGPT-5)

### PR DESCRIPTION
**Pull request by AntoineBastide47**
## Summary
  - Add relationship management system supporting friend requests, friendships, and blocking
  - Implement bidirectional relationship tracking with `RelationshipStatus` enum (Friend, Sent, Pending, Blocked)
  - Create `ProfileScreenViewModel` with validation logic to prevent invalid operations
  - Add comprehensive integration tests (678 lines) covering all relationship workflows

  ## Changes
  - **Account.kt**: Add `relationships` map and helper methods (`isFriendsWith()`, `hasBlocked()`, etc.)
  - **AccountRepository.kt**: Add `sendFriendRequest()`, `acceptFriendRequest()`, `blockUser()`, `resetRelationship()` with atomic Firestore updates
  - **ProfileScreenViewModel.kt**: Implement validation layer preventing invalid operations (duplicate requests, self-operations, wrong-direction accepts)
  - **RelationshipsTests.kt**: Integration tests verifying bidirectional consistency, state transitions, complex workflows, and edge cases

  🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: app/src/main/java/com/github/meeplemeet/model/account/AccountRepository.kt [hunk 1]
```diff
@@ -130,19 +189,133 @@ class AccountRepository : FirestoreRepository("accounts") {
...
-   * The flow emits a complete Account object including all discussion previews. The listeners are
-   * automatically cleaned up when the flow is cancelled.
+   * The listeners are automatically cleaned up when the flow is cancelled.
    *
    * @param accountId The ID of the account to listen to
-   * @return A Flow that emits Account objects when changes are detected
+   * @return A Flow that emits Account objects when changes are detected in the account, its
+   *   previews, or its relationships
    */
   fun listenAccount(accountId: String): Flow<Account> = callbackFlow {
```
Comment (MarcoLourenco55) [inline app/src/main/java/com/github/meeplemeet/model/account/AccountRepository.kt:329]: Consider refactoring this function, due to its high cognitive complexity (https://sonarcloud.io/project/issues?sinceLeakPeriod=true&issueStatuses=OPEN%2CCONFIRMED&pullRequest=193&id=Meeple-Meet_Meeple-Meet&open=AZqxu-r9Fx5fnmI6B-Xq)

File: app/src/main/java/com/github/meeplemeet/model/account/AccountRepository.kt [hunk 2]
```diff
@@ -23,8 +24,75 @@ import kotlinx.coroutines.tasks.await
...
+      collection.document(uid).collection(Account::relationships.name)
+
+  /**
+   * Extracts discussion previews from a Firestore query snapshot.
+   *
+   * @param documents The list of documents from the previews subcollection
+   * @return A map of discussion IDs to their preview objects
+   */
+  private fun extractPreviews(
+      documents: List<com.google.firebase.firestore.DocumentSnapshot>
```
Comment (MarcoLourenco55) [inline app/src/main/java/com/github/meeplemeet/model/account/AccountRepository.kt]: Consider defining imports at the start of the file

File: app/src/main/java/com/github/meeplemeet/model/account/Account.kt [hunk 3]
```diff
@@ -1,24 +1,74 @@
...
+  Sent,
+
+  /** Friend request pending - the other user has sent a request to the current user. */
+  Pending,
+
+  /** Blocked - the current user has blocked the other user. */
+  Blocked,
+
+  /** Deleted - removes the relationship with the other user altogether */
+  Delete,
```
Comment (MarcoLourenco55) [inline app/src/main/java/com/github/meeplemeet/model/account/Account.kt]: I am not sure of understanding the use of this enum here (```Delete```). Since, one user that deletes another one from it's friends list should only have it removed, I don't think this enum is useful per-se. Even in your tests you never use it and only check for presence/absence of relationship

File: app/src/main/java/com/github/meeplemeet/model/account/Account.kt [hunk 4]
```diff
@@ -1,24 +1,74 @@
...
 import kotlinx.serialization.Serializable
 
+/**
+ * Represents the status of a relationship between two users.
+ *
+ * This enum is used to categorize the type of social connection or interaction between the current
+ * user and another user in the system.
+ */
+@Serializable
+enum class RelationshipStatus {
```
Comment (MarcoLourenco55) [inline app/src/main/java/com/github/meeplemeet/model/account/Account.kt:17]: Consider using CAPS for enums (```FRIEND``` instead of ```Friend``` for example)

File: app/src/main/java/com/github/meeplemeet/model/account/Account.kt (hunk 3 above)
Comment (AntoineBastide47) [reply app/src/main/java/com/github/meeplemeet/model/account/Account.kt]: This enum value is used only in the repository to signal that a relationship needs to be removed. It will never be used in the UI and tests. It is just defined as such so that I did not need to add extra parameters to the repository functions.

File: app/src/main/java/com/github/meeplemeet/model/account/Account.kt (hunk 4 above)
Comment (AntoineBastide47) [reply app/src/main/java/com/github/meeplemeet/model/account/Account.kt:17]: I am just using the same style as in NavigationSystem.kt

File: app/src/main/java/com/github/meeplemeet/model/account/ProfileScreenViewModel.kt [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+   * - The accounts are the same user
+   * - Either user has blocked the other (use blockUser to unblock instead)
+   *
+   * If validation passes, the relationship is reset asynchronously via the repository.
+   *
+   * @param account The first account in the relationship
+   * @param friend The second account in the relationship
+   */
+  fun resetRelationship(account: Account, friend: Account) {
+    if (sameOrBlocked(account, friend)) return
```
Comment (MarcoLourenco55) [inline app/src/main/java/com/github/meeplemeet/model/account/ProfileScreenViewModel.kt:155]: Following the Kdoc of this function it would be possible to unblock someone by resetting the relationship
I may have understood it incorrectly, but this line seems to make this impossible, due to the early returning

Your tests also assert for this impossibility (in ```fun viewModelResetRelationship_preventsInvalidOperations```)

Make sure the viewmodel follows the same principle than the repository

File: app/src/main/java/com/github/meeplemeet/model/account/Account.kt [hunk 6]
```diff
@@ -29,13 +79,25 @@ data class Account(
...
+    val relationships: Map<String, RelationshipStatus> = emptyMap()
 )
 
 /**
  * Minimal serializable form of [Account] without the UID, used for Firestore storage.
  *
- * Firestore stores the UID as the document ID, so it is omitted from the stored object.
+ * Firestore stores the UID as the document ID, so it is omitted from the stored object. This class
+ * is optimized for serialization and contains only the fields that need to be persisted in the
+ * database. The relationships are stored as a flat list rather than separate fields for friends,
```
Comment (MarcoLourenco55) [inline app/src/main/java/com/github/meeplemeet/model/account/Account.kt]: Maybe I missunderstood what you meant by "flat list", but is this Kdoc up to date with the new version of the code ?

File: app/src/main/java/com/github/meeplemeet/model/account/ProfileScreenViewModel.kt (hunk 5 above)
Comment (AntoineBastide47) [reply app/src/main/java/com/github/meeplemeet/model/account/ProfileScreenViewModel.kt:155]: Ah yes good catch

File: app/src/main/java/com/github/meeplemeet/model/account/Account.kt [hunk 7]
```diff
@@ -51,24 +113,40 @@ data class AccountNoUid(
...
-        photoUrl = accountNoUid.photoUrl,
-        description = accountNoUid.description,
-        shopOwner = accountNoUid.shopOwner,
-        spaceRenter = accountNoUid.spaceRenter,
-    )
+    previews: Map<String, DiscussionPreviewNoUid> = emptyMap(),
+    relationships: List<Relationship> = emptyList()
+): Account {
+  // Convert relationship list to map for efficient lookup
+  val mappedRelationships = mutableMapOf<String, RelationshipStatus>()
```
Comment (MarcoLourenco55) [inline app/src/main/java/com/github/meeplemeet/model/account/Account.kt]: Consider using a functional approach here, since it is cleaner and safer : ```val mappedRelationships = relationships.associate { it.uid to it.status }```

File: app/src/main/java/com/github/meeplemeet/model/account/AccountRepository.kt [hunk 8]
```diff
@@ -49,51 +117,42 @@ class AccountRepository : FirestoreRepository("accounts") {
...
    * @param id The account ID to retrieve
-   * @return The Account object with populated discussion previews
+   * @return The Account object with populated discussion previews and relationships map
    * @throws AccountNotFoundException if the account does not exist
    */
   suspend fun getAccount(id: String): Account {
     val snapshot = collection.document(id).get().await()
     val account = snapshot.toObject(AccountNoUid::class.java) ?: throw AccountNotFoundException()
 
     val previewsSnap = collection.document(id).collection("previews").get().await()
```
Comment (MarcoLourenco55) [inline app/src/main/java/com/github/meeplemeet/model/account/AccountRepository.kt:133]: I think replacing ```"previews``` with ```Account::previews.name``` may be more consistent with your other collection calls

File: app/src/main/java/com/github/meeplemeet/model/account/Account.kt (hunk 4 above)
Comment (MarcoLourenco55) [reply app/src/main/java/com/github/meeplemeet/model/account/Account.kt:17]: I think the idiomatic way of calling it in Kotlin is using CAPS. I don't think duplicating non-idiomatic writing is a good idea, even for consistency across the project. I think it is better renaming those used elsewhere.

File: app/src/main/java/com/github/meeplemeet/model/account/AccountRepository.kt [hunk 9]
```diff
@@ -130,53 +188,295 @@ class AccountRepository : FirestoreRepository("accounts") {
...
+   *   performing the blocking action
+   */
+  suspend fun blockUser(accountId: String, otherId: String, otherBlockedAccount: Boolean) {
+    val batch = db.batch()
+
+    val aRef = relationships(accountId).document(otherId)
+    val bRef = relationships(otherId).document(accountId)
+
+    batch[aRef] = mapOf(FIELD_STATUS to RelationshipStatus.Blocked)
+    if (!otherBlockedAccount) batch.delete(bRef)
```
Comment (MarcoLourenco55) [inline app/src/main/java/com/github/meeplemeet/model/account/AccountRepository.kt]: I think this may create a race condition on blocking status from both persons, if done at the same time

File: app/src/main/java/com/github/meeplemeet/model/account/AccountRepository.kt [hunk 10]
```diff
@@ -130,53 +188,295 @@ class AccountRepository : FirestoreRepository("accounts") {
...
+     * @return A ListenerRegistration that can be used to remove the listener
+     */
+    private fun listenToPreviews(
+        accountRef: com.google.firebase.firestore.DocumentReference,
+        flow: kotlinx.coroutines.channels.ProducerScope<Account>,
+        accountId: String,
+        accountNoUid: AccountNoUid
+    ): ListenerRegistration {
+      return accountRef.collection(Account::previews.name).addSnapshotListener { qs, e ->
+        if (e != null) {
```
Comment (MarcoLourenco55) [inline app/src/main/java/com/github/meeplemeet/model/account/AccountRepository.kt:401]: Consider logging error information here and on the listening functions below

File: app/src/main/java/com/github/meeplemeet/model/account/AccountRepository.kt (hunk 8 above)
Comment (AntoineBastide47) [reply app/src/main/java/com/github/meeplemeet/model/account/AccountRepository.kt:133]: This change has been made on the `feat/notification-backend` that will be in a PR very soon, so I won't address it here.

File: app/src/main/java/com/github/meeplemeet/model/account/Account.kt (hunk 6 above)
Comment (AntoineBastide47) [reply app/src/main/java/com/github/meeplemeet/model/account/Account.kt]: Fixed

File: app/src/main/java/com/github/meeplemeet/model/account/Account.kt (hunk 7 above)
Comment (AntoineBastide47) [reply app/src/main/java/com/github/meeplemeet/model/account/Account.kt]: Done

File: app/src/main/java/com/github/meeplemeet/model/account/AccountRepository.kt (hunk 9 above)
Comment (AntoineBastide47) [reply app/src/main/java/com/github/meeplemeet/model/account/AccountRepository.kt]: I changed it to a transaction instead of a batch which will prevent the race condition

File: app/src/main/java/com/github/meeplemeet/model/account/AccountRepository.kt (hunk 10 above)
Comment (AntoineBastide47) [reply app/src/main/java/com/github/meeplemeet/model/account/AccountRepository.kt:401]: This isn't a real "error" it only fails if either there is no connection or there are no previews. So this isn't an error since without internet, we can't do anything so return nothing and same for when there are no previews.

File: app/src/main/java/com/github/meeplemeet/model/account/Account.kt (hunk 3 above)
Comment (MarcoLourenco55) [reply app/src/main/java/com/github/meeplemeet/model/account/Account.kt]: I see your point, but I think enums should declare "states" and not "operations"

I think it would be clearer to have it removed and use ```null``` instead in ```setFriendStatus``` and ```resetRelationship```. You can make use of ```?``` type definition to do so

**Review by MarcoLourenco55 — CHANGES_REQUESTED** [review]
Thank you very much for introducing the friend's management backend, it will make the app much more usable !

A part from the comments and requests done in specific parts, here are some general comments that could improve the code:

*  Consider using more readable variables inside your functions, such that a reader can understand it directly. For example, in ```blockUser``` function you refer to the relationship status as ```a```, which is not human friendly. But other examples exists in the code.

* Consider making more concise and clearer commits. For example, commit https://github.com/Meeple-Meet/Meeple-Meet/pull/193/commits/b701426ca19d3411631aaa3b96a60355ed97768c, could be separated in several small ones (each solving one of the comments in this PR). Moreover, the descriptions should contain information such that one can understand quite easily what you have done

* (see comment above) Consider attaching these commits to their respective comment, such that the reviewer can easily see that it was solved and check the changes. You can find  examples in the following PRs : https://github.com/Meeple-Meet/Meeple-Meet/pull/125 and https://github.com/Meeple-Meet/Meeple-Meet/pull/162

File: app/src/main/java/com/github/meeplemeet/model/account/Account.kt (hunk 3 above)
Comment (AntoineBastide47) [reply app/src/main/java/com/github/meeplemeet/model/account/Account.kt]: Fixed in b125badd92b3ec8d2240acb69655b676bc13a191

File: app/src/main/java/com/github/meeplemeet/model/account/Account.kt (hunk 4 above)
Comment (AntoineBastide47) [reply app/src/main/java/com/github/meeplemeet/model/account/Account.kt:17]: Fixed in b125badd92b3ec8d2240acb69655b676bc13a191

**Review by MarcoLourenco55 — CHANGES_REQUESTED** [review]
Thank you for the functionality changes, I think the code is clearer now.
However, as already commented before, I think that renaming some of your variables to make them more human-readable is necessary. Please consider applying this change prior to merge.

A part from that, the code LGTM !

**Review by MarcoLourenco55 — CHANGES_REQUESTED** [review]
There are still some places where renaming is necessary, as for example : ```blockUser```,  ```setFriendStatus```, ```listenToPreviews``` and ```listenToRelationships```

Example of such variables that could be better named are:
* ```aRef``` and ```bRef```
* ```tx```
* ```qs```
* ```e``` (this one could be discussed)

**Review by MarcoLourenco55 — CHANGES_REQUESTED** [review]
Thank you for renaming some of these variables, but I think some of them could be renamed too. I guess they make sense for you, but for a new reader, they may not be clear enough in my opinion

Also consider naming more precisely your commits. I think something like "renaming variables" would be greater for example for this commit https://github.com/Meeple-Meet/Meeple-Meet/pull/193/commits/2f3b6d8f58ee75120becad4eddfc3d6c823baf54

**Comment by AntoineBastide47** [comment]
`qs` is an abreviation for `QuerySnapshot` as that is the type returned by the listeners.
`tx` is an abreviation for `Transaction`.
`eN` is for `errorN`

So the naming is correct.

**Comment by MarcoLourenco55** [comment]
I think renaming them to ```querySnap``` and ```transaction``` could be better than ```qs``` and ```tx```

Of course, renaming are always prune to reader comprehension. For example, I didn't pointed out about the ```db``` one (abbrevation for data base), since for me this is quite clear from the code and the context for a first time reader.

Maybe ```qs``` and ```tx``` are fine too, but I had to dig on the code to be able to understand them. I could not do this at first sight.

I think we should ask the opinion on a third member such that he can decide on which version he/she prefers. What do you think @AlexandroHattoum ?

**Comment by AlexandroHattoum** [comment]
> I think renaming them to `querySnap` and `transaction` could be better than `qs` and `tx`
> 
> Of course, renaming are always prune to reader comprehension. For example, I didn't pointed out about the `db` one (abbrevation for data base), since for me this is quite clear from the code and the context for a first time reader.
> 
> Maybe `qs` and `tx` are fine too, but I had to dig on the code to be able to understand them. I could not do this at first sight.
> 
> I think we should ask the opinion on a third member such that he can decide on which version he/she prefers. What do you think @AlexandroHattoum ?

They're both clear enough as abbreviations. I think you have to anyways look at the variable declaration, and that will give you a good enough idea of what you're using/what you're dealing with. If it's unclear to others just add a short comment above/ to the right of the line.

**Review by MarcoLourenco55 — APPROVED** [review]
Since the last two renaming seem fine to @AlexandroHattoum, I have nothing more to say about the rest of the code that LGTM !

Thank you again for introducing the friend's management backend
